core/coreunix/add.go

```diff
@@ -241,96 +329,111 @@ func (params *Adder) addNode(node *dag.Node, path string) error {
 		path = key.Pretty()
 	}
 
-	if err := params.editor.InsertNodeAtPath(params.ctx, path, node, newDirNode); err != nil {
+	if err := mfs.PutNode(params.mr, path, node); err != nil {
 		return err
 	}
 
-	return outputDagnode(params.out, path, node)
+	if !params.Silent {
+		return outputDagnode(params.out, path, node)
+	}
+	return nil
 }
 
 // Add the given file while respecting the params.
-func (params *Adder) AddFile(file files.File) (*dag.Node, error) {
+func (params *Adder) AddFile(file files.File) error {
+	params.unlock = params.node.Blockstore.PinLock()
+	defer params.unlock()
+
+	return params.addFile(file)
+}
+
+func (adder *Adder) addFile(file files.File) error {
+	err := adder.maybePauseForGC()
+	if err != nil {
+		return err
+	}
+
 	switch {
-	case files.IsHidden(file) && !params.Hidden:
+	case files.IsHidden(file) && !adder.Hidden:
 		log.Debugf("%s is hidden, skipping", file.FileName())
-		return nil, &hiddenFileError{file.FileName()}
+		return &hiddenFileError{file.FileName()}
 	case file.IsDirectory():
-		return params.addDir(file)
+		return adder.addDir(file)
 	}
 
 	// case for symlink
 	if s, ok := file.(*files.Symlink); ok {
 		sdata, err := unixfs.SymlinkData(s.Target)
 		if err != nil {
-			return nil, err
+			return err
 		}
 
 		dagnode := &dag.Node{Data: sdata}
-		_, err = params.node.DAG.Add(dagnode)
+		_, err = adder.node.DAG.Add(dagnode)
 		if err != nil {
-			return nil, err
+			return err
 		}
 
-		err = params.addNode(dagnode, s.FileName())
-		return dagnode, err
+		return adder.addNode(dagnode, s.FileName())
 	}
 
 	// case for regular file
 	// if the progress flag was specified, wrap the file so that we can send
 	// progress updates to the client (over the output channel)
 	var reader io.Reader = file
-	if params.Progress {
-		reader = &progressReader{file: file, out: params.out}
+	if adder.Progress {
+		reader = &progressReader{file: file, out: adder.out}
 	}
 
-	dagnode, err := params.add(reader)
+	dagnode, err := adder.add(reader)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	// patch it into the root
-	log.Infof("adding file: %s", file.FileName())
-	err = params.addNode(dagnode, file.FileName())
-	return dagnode, err
+	return adder.addNode(dagnode, file.FileName())
 }
 
-func (params *Adder) addDir(dir files.File) (*dag.Node, error) {
-	tree := newDirNode()
+func (params *Adder) addDir(dir files.File) error {
 	log.Infof("adding directory: %s", dir.FileName())
 
+	err := mfs.Mkdir(params.mr, dir.FileName(), true)
+	if err != nil {
+		return err
+	}
+
 	for {
 		file, err := dir.NextFile()
 		if err != nil && err != io.EOF {
-			return nil, err
+			return err
 		}
 		if file == nil {
 			break
 		}
 
-		node, err := params.AddFile(file)
+		err = params.addFile(file)
```

Comment:
ok so i think directories should add every file in a new goroutine (with rate limiting). i had started on this earlier, but aborted. the general idea was every dir -> files fan out uses goroutines, like:

```go
ctx, cancel := context.WithCancel(params.ctx)
defer cancel()
// this context will cancel out all the children if we error out early.
// would need to wire the addFile funcs with a ctx. OR we _could_ cancel 
// the entire context of thea adder instead...

errs := make(chan error, 10) // some room to avoid waiting for errors
                // dont think we have access to read len(dir) though

for {
  file, err := dir.NextFile()
  if err != nil && err != io.EOF {
    return err // defer cancel unblocks other children
  }
  if file == nil {
    break // done with files
  }

  <-params.concurrency // adder-global parameter
  go func(file) {
    defer func() {
      params.concurrency<- struct{}{} // done
    }()

    err := params.addFile(ctx, file) // ctx cancel should unblock addFile

    // return even nil errors, so we force the parent to wait
    // until _all_ children are done during non-error ops.
    // but bail on ctx.Done.
    select {
    case errs<- err:
    case ctx.Done():
    }
  }(file)
}

for err := range errs {
  if err == nil {
    continue
  }
  if _, ok := err.(*hiddenFileError); ok {
    continue // hidden file error, skip file
  }
  // error!
  return err // defer cancel unblocks + terminates all children.
}

// all children done.
return nil
```